source/common/common/trie.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+/**
+ * This is a general purpose trie. Keys and Values can be arbitrary classes
+ * provided they can be hashed. It supports both partial matches and exact
+ * matches.
+ *
+ * Note that Value must have a conversion constructor that will take nullptr.
+ */
+// TODO(tschroed): It's possible for Key to be a different type from the
+// internal path component. E.g. a uint32_t IPv4 prefix can be tokenized to
+// uint8_t components. Likewise a string representation of same.
+template <class Key, class Value, class Tokenizer> class TrieNode {
+public:
+  TrieNode(Value value) : value_(value) {}
+  TrieNode() : TrieNode(nullptr) {}
+  ~TrieNode() {}
+
+  /**
+   * Insert a value into the trie.
+   * @param key the name associated with the value.
+   * @param value the value to store on the node associated with name
+   */
+  void emplace(const Key& key, Value value);
+
+  /**
+   * Lookup a value by key.
+   * @param key the path name to look up
+   * @return a std::pair consisting of the value if found and a boolean
+   *         indicating if the match was exact.
+   */
+  std::pair<Value, bool> find(const Key& key) const;
+
+private:
+  std::pair<Value, bool> find(std::vector<Key>& path_components) const;
+  void emplace(std::vector<Key>& path_components, Value value);
+  Value value() const { return value_; }
+  void set_value(Value value) { value_ = value; }
+
+  // By holding pointers to the TriedNodes we don't have to worry that they're
+  // incomplete types.
+  std::unordered_map<Key, std::unique_ptr<TrieNode<Key, Value, Tokenizer>>> children_;
+  Value value_;
+  Tokenizer tokenizer_;
+};
+
+template <class Key, class Value, class Tokenizer>
+void TrieNode<Key, Value, Tokenizer>::emplace(const Key& key, Value value) {
+  std::vector<Key> path_components = tokenizer_.tokenize(key);
+  // Keys come in big endian but we reverse them for vector iteration
```

Comment:
This comment is a little strange regarding endian, if I understand you reverse here so that you can pop_back later, I would just say that.